kmod/src/forest.c

```diff
@@ -801,23 +810,162 @@ int scoutfs_forest_oino_alloc(struct super_block *sb, u64 trans_seq)
 	return ret;
 }
 
-void scoutfs_forest_oino_update(struct super_block *sb)
+int scoutfs_forest_oino_update(struct super_block *sb)
 {
 	struct scoutfs_bloom_block *bb;
 	DECLARE_FOREST_INFO(sb, finf);
+	u64 last_seq;
+	int ret;
 	int i;
 
 	if (!finf->oino_bl)
-		return;
+		return 0;
 
 	bb = finf->oino_bl->data;
 
+	ret = scoutfs_client_get_last_seq(sb, &last_seq);
+	WARN_ON_ONCE(ret);
+
+	bb->hdr.seq = cpu_to_le64(last_seq);
+	finf->our_log.oino_bloom_ref.seq = bb->hdr.seq;
+
 	/* Turn the counted BF into a classic BF */
 	for (i = 0; i < SCOUTFS_FOREST_BLOOM_BITS; i++) {
 		if (atomic_read(&finf->oino_counted_bloom[i]))
 			set_bit_le(i, bb->bits);
 	}
-
 	scoutfs_block_put(sb, finf->oino_bl);
 	finf->oino_bl = NULL;
+
+	schedule_delayed_work(&finf->oino_work, msecs_to_jiffies(OINO_WORK_DELAY));
+
+	return 0;
+}
+
+void scoutfs_oino_delete_work(struct work_struct *work)
+{
+	struct forest_info *finf = container_of(work, struct forest_info,
+						oino_work.work);
+	struct super_block *sb = finf->sb;
+	struct scoutfs_net_roots roots;
+	SCOUTFS_BTREE_ITEM_REF(iref);
+	struct scoutfs_log_trees lt = {{{ 0 }}};
+	struct scoutfs_bloom_block *bb;
+	struct scoutfs_bloom_block *first_bb = NULL;
+	struct scoutfs_block *bl;
+	struct scoutfs_block *first_bl = NULL;
+	struct scoutfs_key ltk;
+	int ret = 0;
+	int i;
+
+	if (finf->shutdown)
+		return;
+
+	scoutfs_inc_counter(sb, forest_oino_delete_work);
+
+	ret = scoutfs_client_get_roots(sb, &roots);
+	if (ret)
+		goto out;
+
+	scoutfs_key_init_log_trees(&ltk, 0, 0);
+	for (;; scoutfs_key_inc(&ltk)) {
+		ret = scoutfs_btree_next(sb, &roots.logs_root, &ltk, &iref);
+		if (ret == 0) {
+			if (iref.val_len == sizeof(lt)) {
+				ltk = *iref.key;
+				memcpy(&lt, iref.val, sizeof(lt));
+			} else {
+				ret = -EIO;
+			}
+			scoutfs_btree_put_iref(&iref);
+		}
+		if (ret < 0) {
+			if (ret == -ENOENT)
+				break;
+			goto out; /* including stale */
+		}
+
+		if (lt.oino_bloom_ref.blkno == 0)
```

Comment:
That is a little annoying. I think I proposed something before and it was bad for a reason that I neglected to write down alas --> why not ask the server for # of clients or even a list of rids of all clients and use that to abort if # of blooms is too few/don't match?

Reply:
The number of clients isn't interesting, that doesn't tell you anything about the relative ages of the clients or the files they may have open or not. And the list of rids doesn't help either.   If you have all the rids, but one of them has blkno == 0, then we're back to where we started.

The problem is that blkno == 0 both when 1) you've first mounted and have opened things but haven't committed anything yet, and 2) you've been evicted and the oino bloom block has been freed.

In #1 we don't know *when* they mounted so we don't know if they were able to open files that have since been unlinked.  In #2 they're gone and we don't care and can ignore them.

I think the solution to all this is to create a valid empty oino bloom block with the client's first log_trees item that is created by the server when get_trees doesn't find one.  This is done early enough in mount that the client won't have any files open yet.  If we initialize last_seq in the initial oino bloom then we know that clients always have an oino bloom block in their log_trees item and we don't have to do any special nonsense when deciding if it's safe to delete.  Just iterate over the items and merge what we see.

Reply:
OK I'll give that a go.